src/libraries/Common/src/System/Net/IPv4AddressHelper.Common.cs

```diff
@@ -105,32 +107,48 @@ internal static unsafe bool IsValid(char* name, int start, ref int end, bool all
         //                 / "2" %x30-34 DIGIT     ; 200-249
         //                 / "25" %x30-35          ; 250-255
         //
-        internal static unsafe bool IsValidCanonical(char* name, int start, ref int end, bool allowIPv6, bool notImplicitFile)
+        internal static bool IsValidCanonical<TChar>(ReadOnlySpan<TChar> name, out int bytesConsumed, bool allowIPv6, bool notImplicitFile)
+            where TChar : unmanaged, IBinaryInteger<TChar>
         {
+            TChar ComponentSeparator = TChar.CreateChecked('.');
+            ReadOnlySpan<TChar> PrefixSeparators = [TChar.CreateChecked('/'), TChar.CreateChecked('\\')];
+            ReadOnlySpan<TChar> UrlSeparators = [TChar.CreateChecked(':'), TChar.CreateChecked('?'), TChar.CreateChecked('#')];
+            TChar OctalPrefix = TChar.CreateChecked('0');
+
+            ReadOnlySpan<TChar> IPv6Terminators = [TChar.CreateChecked('%'), TChar.CreateChecked('/'), TChar.CreateChecked(']')];
```

Comment:
Storing this in Spans instead of inlining the values like they were inlined before creates extra work for the code generator.

I am pretty sure that this is going to be a performance regression for Mono that has less capable optimizer. It may be a performance regression for RyuJIT too.

It would be best to inline the values at the point of use like they were inlined before. You will have hard time avoiding performance regressions otherwise.

Reply:
Thanks, I didn't realise that. I've inlined the Spans (along with the other variables @MihaZupan - this also resolves your comment there too.) Where the purpose of some of the characters wasn't immediately obvious, I've also added comments.